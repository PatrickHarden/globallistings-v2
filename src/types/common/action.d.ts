import { Action } from 'redux';

export interface ActionPayload<T> {
  readonly payload: T;
}

export interface ActionMeta<T> {
  readonly meta: T;
}

export interface ReceivedAt {
  receivedAt: number;
}

export interface HashMeta {
  readonly hash: string;
}

export interface ErrorPayload {
  readonly payload: Error;
}

export type SuccessAction<T> = Action<string> &
  ActionPayload<T> &
  ActionMeta<HashMeta & ReceivedAt>;

export type FailureAction = Action & ActionMeta<HashMeta> & ErrorPayload;
export type RequestAction = Action & ActionMeta<HashMeta>;
