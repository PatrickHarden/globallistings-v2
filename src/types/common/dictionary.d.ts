export type Dictionary<T> = { [key: string]: T | undefined };
export type ExplicitDictionary<T> = { [key: string]: T };
