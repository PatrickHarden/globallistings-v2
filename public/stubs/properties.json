{
  "Found": true,
  "DocumentCount": 69,
  "Documents": [
    [
      {
        "Common.ActualAddress": {
          "Common.Line1": "Tynemouth House",
          "Common.Line2": "Benton Park View",
          "Common.Country": "GB",
          "Common.PostCode": "NE98 1ZZ"
        },
        "Common.Agency": {
          "Common.AgentAddress": {
            "Common.Line1": "Canary Wharf",
            "Common.Line2": "Millennium Harbour",
            "Common.Line3": "22 Westferry Road",
            "Common.Locallity": "London",
            "Common.Country": "GB",
            "Common.PostCode": "E14 8LW"
          },
          "Common.AgencyName": "Canary Wharf Sales",
          "Common.EmailAddress": "residential.canarywharf@cbre.com",
          "Common.TelephoneNumber": "020 7519 5900"
        },
        "Common.Agents": [
          {
            "Common.EmailAddress": "darren.gibbinson@cbre.com",
            "Common.AgentName": "Darren Gibbinson",
            "Common.TelephoneNumber": "020 7519 5920",
            "Common.Website": "http://www.cbre.com/ddddyyy.html"
          },
          {
            "Common.EmailAddress": "david.stephenson@cbre.com",
            "Common.AgentName": "David Stephenson",
            "Common.TelephoneNumber": "020 9999 5920",
            "Common.Website": "http://www.cbre.com/ddddyyy.html"
          }
        ],
        "Common.ContactGroup": {
          "Common.GroupName": [
            {
              "Common.CultureCode": "en-GB",
              "Common.Text": "West End Sales Office"
            },
            {
              "Common.CultureCode": "zh-CN",
              "Common.Text": "西区销售办事处"
            }
          ],
          "Common.Address": {
            "Common.Line1": "Canary Wharf",
            "Common.Line2": "Millennium Harbour",
            "Common.Line3": "22 Westferry Road",
            "Common.Locallity": "London",
            "Common.Country": "GB",
            "Common.PostCode": "E14 8LW"
          },
          "Common.GroupType": "AreaOffice",
          "Common.Website": "http://www.cbre.com/ddddyyy.html",
          "Common.Contacts": [
            {
              "Common.EmailAddress": "darren.gibbinson@cbre.com",
              "Common.TelephoneNumber": "020 7519 5920"
            },
            {
              "Common.EmailAddress": "david.stephenson@cbre.com",
              "Common.AgentName": "David Stephenson",
              "Common.TelephoneNumber": "020 9999 5920"
            }
          ]
        },
        "Common.AvailableFrom": "2015-08-09",
        "Common.Brochures": [
          {
            "Common.Uri":
              "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/Brochure/2_278296/Tempest%20eBrochure.pdf",
            "Common.Culture": "en-GB"
          },
          {
            "Common.Uri":
              "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/Brochure/2_278296/Tempest%20eBrochure01.pdf",
            "Common.Culture": "en-GB"
          },
          {
            "Common.Uri":
              "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/Brochure/2_278296/Tithebarn%20Project%20Leisure%20brochure%20Final.pdf",
            "Common.Culture": "zh-CN"
          }
        ],
        "Common.Charges": [
          {
            "Common.ChargeKind": "BusinessRates",
            "Common.CurrencyCode": "GBP",
            "Common.Interval": "Monthly",
            "Common.Amount": 499,
            "Common.OnApplication": false,
            "Common.Year": 2014,
            "Common.PerUnit": "PerSqFt",
            "Common.Exact": true,
            "Common.TaxModifer": "PlusVAT"
          },
          {
            "Common.ChargeKind": "Rent",
            "Common.CurrencyCode": "GBP",
            "Common.Interval": "Monthly",
            "Common.Amount": 4534,
            "Common.PerUnit": "Whole",
            "Common.Exact": true,
            "Common.TaxModifer": "PlusSalesTax"
          },
          {
            "Common.ChargeKind": "EstateCharge",
            "Common.CurrencyCode": "GBP",
            "Common.Interval": "Annually",
            "Common.Amount": 12000,
            "Common.Year": 2014,
            "Common.PerUnit": "Whole",
            "Common.Exact": true,
            "Common.TaxModifer": "IncludingVAT"
          },
          {
            "Common.ChargeKind": "EstateCharge",
            "Common.CurrencyCode": "USD",
            "Common.Interval": "Weekly",
            "Common.Amount": 12000,
            "Common.Year": 2014,
            "Common.PerUnit": "Whole",
            "Common.Exact": true,
            "Common.TaxModifer": "IncludingSalesTax "
          },
          {
            "Common.ChargeKind": "SalePrice",
            "Common.CurrencyCode": "GBP",
            "Common.Interval": "Once",
            "Common.Amount": 1850000,
            "Common.PerUnit": "Whole",
            "Common.Exact": false
          },
          {
            "Common.ChargeKind": "ServiceCharge",
            "Common.CurrencyCode": "GBP",
            "Common.Interval": "Monthly",
            "Common.Amount": 299,
            "Common.Year": 2014,
            "Common.PerUnit": "PerSqFt"
          }
        ],
        "Common.Coordinate": {
          "lat": 51.518065,
          "lon": -0.117154
        },
        "Common.Created": "2015-07-01T10:32:14",
        "Common.FloorPlans": [
          {
            "Common.ImageCaption": "Image.Photo-1",
            "Common.ImageResources": [
              {
                "Common.Image.Height": 1920,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/floorplans/cbrrps-CAR150038_FloorPlan_1_1280x960.png",
                "Common.Image.Width": 640
              }
            ]
          },
          {
            "Common.ImageCaption": "Image.Photo-2",
            "Common.ImageResources": [
              {
                "Common.Image.Height": 1920,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/floorplans/cbrrps-CAR150038_FloorPlan_2_1280x960.png",
                "Common.Image.Width": 640
              }
            ]
          }
        ],
        "Common.FloorsAndUnits": [
          {
            "Common.SubdivisionName": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "Atrium"
              }
            ],
            "Common.Areas": [
              {
                "Common.Units": "sqft",
                "Common.Area": 3702
              },
              {
                "Common.Units": "sqm",
                "Common.Area": 344
              }
            ],
            "Common.Unit.Status": "Availible",
            "Common.Unit.Use": "Office",
            "Common.AvailableFrom": "2015-10-09",
            "Common.Charges": [
              {
                "Common.ChargeKind": "SalePrice",
                "Common.CurrencyCode": "GBP",
                "Common.Interval": "Once",
                "Common.Amount": 20000,
                "Common.PerUnit": "Whole",
                "Common.Exact": true
              }
            ]
          },
          {
            "Common.SubdivisionName": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "Reception"
              }
            ],
            "Common.Areas": [
              {
                "Common.Units": "sqft",
                "Common.Area": 1776
              }
            ],
            "Common.Unit.Status": "Availible",
            "Common.Unit.Use": "Office"
          },
          {
            "Common.SubdivisionName": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "Gnd"
              }
            ],
            "Common.Areas": [
              {
                "Common.Units": "sqft",
                "Common.Area": 9091
              },
              {
                "Common.Units": "sqm",
                "Common.Area": 909
              }
            ],
            "Common.Unit.Status": "Availible",
            "Common.Unit.Use": "Office",
            "Common.AvailableFrom": "2015-10-09",
            "Common.Charges": [
              {
                "Common.ChargeKind": "SalePrice",
                "Common.CurrencyCode": "GBP",
                "Common.Interval": "Once",
                "Common.Amount": 20000,
                "Common.PerUnit": "Whole",
                "Common.Exact": true
              }
            ]
          }
        ],
        "Common.HomeSite": "uk-resi",
        "Common.Highlights": [
          {
            "Common.Highlight": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "Air Conditioning"
              },
              {
                "Common.CultureCode": "es-ES",
                "Common.Text": "Aire Acondicionado"
              }
            ]
          },
          {
            "Common.Highlight": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "Excellent BREEAM  Rating"
              }
            ]
          },
          {
            "Common.Highlight": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "B Rating EPC"
              }
            ]
          },
          {
            "Common.Highlight": [
              {
                "Common.CultureCode": "en-GB",
                "Common.Text": "1:797 sq ft Parking Ratio"
              }
            ]
          }
        ],
        "Common.LandSize": [
          {
            "Common.Units": "acre",
            "Common.Area": 3702
          },
          {
            "Common.Units": "hectare",
            "Common.Area": 344
          }
        ],
        "Common.LastProcessed": "2015-07-01T10:35:34.032",
        "Common.LastUpdated": "2015-07-01T10:35:34.032",
        "Common.SourceLastupdated": "2015-07-01T09:35:34.032",
        "Common.LeaseTypes": ["Groundease", "OccupationalLease"],
        "Common.LongDescription": [
          {
            "Common.CultureCode": "en-GB",
            "Common.Text":
              "This well-proportioned two bedroom apartment is centrally located in the heart of Holborn. The apartment is situated on the third floor of a popular modern residential block with the benefit of lift access, CCTV surveillance in the entrance hallway and a communal bicycle park."
          },
          {
            "Common.CultureCode": "zh-CN",
            "Common.Text":
              "这匀称的两居室公寓坐落在霍尔本的心脏。该公寓坐落在与电梯访问的利益，闭路电视监控，在入口处走廊和公共自行车公园流行的现代住宅楼的三楼。"
          }
        ],
        "Common.MaximumSize": [
          {
            "Common.Units": "sqft",
            "Common.Area": 3702
          },
          {
            "Common.Units": "sqm",
            "Common.Area": 344
          }
        ],
        "Common.MinimumSize": [
          {
            "Common.Units": "sqft",
            "Common.Area": 370
          },
          {
            "Common.Units": "sqm",
            "Common.Area": 34
          }
        ],
        "Common.NumberOfBedrooms": 3,
        "Common.Photos": [
          {
            "Common.ImageCaption": "Image.Photo-1",
            "Common.ImageResources": [
              {
                "Common.Breakpoint": "Small",
                "Common.Image.Height": 480,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_1_640x480.jpg",
                "Common.Image.Width": 640
              },
              {
                "Common.Breakpoint": "Medium",
                "Common.Image.Height": 960,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_1_1280x960.jpg",
                "Common.Image.Width": 1280
              },
              {
                "Common.Breakpoint": "Large",
                "Common.Image.Height": 1920,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_1_2560x1920.jpg",
                "Common.Image.Width": 2560
              }
            ]
          },
          {
            "Common.ImageCaption": "Image.Photo-2",
            "Common.ImageResources": [
              {
                "Common.Image.Height": 480,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_2_640x480.jpg",
                "Common.Image.Width": 640
              },
              {
                "Common.Image.Height": 960,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_2_1280x960.jpg",
                "Common.Image.Width": 1280
              },
              {
                "Common.Image.Height": 1920,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_2_2560x1920.jpg",
                "Common.Image.Width": 2560
              }
            ]
          },
          {
            "Common.ImageCaption": "Image.Photo-3",
            "Common.ImageResources": [
              {
                "Common.Image.Height": 480,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_3_640x480.jpg",
                "Common.Image.Width": 640
              },
              {
                "Common.Image.Height": 960,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_3_1280x960.jpg",
                "Common.Image.Width": 1280
              },
              {
                "Common.Image.Height": 1920,
                "Common.Resource.Uri":
                  "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/photos/cbrrps-BOW120225_Photo_3_2560x1920.jpg",
                "Common.Image.Width": 2560
              }
            ]
          }
        ],
        "Common.PrimaryKey": "GB_Plus_27875",
        "Common.PropertyTypes": ["Apartment"],
        "Common.RestrictedSites": ["sg-resi", "ca-resi"],
        "Common.Strapline": [
          {
            "Common.CultureCode": "en-GB",
            "Common.Text": "Fashionable living close to Canary Wharf centre"
          },
          {
            "Common.CultureCode": "zh-CN",
            "Common.Text": "时尚生活靠近金丝雀码头中心"
          }
        ],
        "Common.TotalSize": [
          {
            "Common.Units": "sqft",
            "Common.Area": 185940.0
          }
        ],
        "Common.EnergyPerformanceInformation": [
          {
            "Common.Uri":
              "file://emea.cbre.net/Data/WebPlatform/Listing Service/PRD/Agency365Data/OfficePLUS/EPC/cbrrps-CAR150038_EPC_1_1280x960.png",
            "Common.Culture": "en-GB"
          }
        ],
        "Common.UsageType": "Residential",
        "Common.Website": "http://www.cbre.com/someproperty.html",
        "Common.Walkthrough":
          "http://api.inspiredlabs.com/viewer?embed=247700361404341",
        "Common.ParentProperty": "GB_Plus_17344",
        "UnitedKingdom.VATPayable": true,
        "Dynamic.SalePrice": 1600000,
        "Dynamic.LettingChargeAnnually": "0.077",
        "Dynamic.LettingChargeWeekly": "0.0777",
        "Dynamic.LettingChargeMonthly": "0.07777",
        "Dynamic.MinArea": "7000.0007",
        "Dynamic.MaxArea": "7700.7007",
        "Dynamic.NewHome": false,
        "Dynamic.UnderOffer": false,
        "Dynamic.LetUnderOffer": false
      }
    ]
  ],
  "Aggregations": [[]],
  "Took": "00:00:00.0246442"
}
