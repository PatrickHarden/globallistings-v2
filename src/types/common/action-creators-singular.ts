export const createRequestFailure = <T extends string>(type: T) => (payload: Error) => ({
  type,
  payload,
});

export const createRequestSuccess = <P, T extends string>(type: T) => (payload: P) => ({
  type,
  payload,
  meta: {
    receivedAt: Date.now(),
  },
});

export const createRequestPending = <T extends string>(type: T) => () => ({
  type,
});
