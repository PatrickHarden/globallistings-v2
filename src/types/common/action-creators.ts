export const createRequestFailure = <T extends string>(type: T) => (
  payload: Error,
  hash: string
) => ({
  type,
  payload,
  meta: {
    hash,
  },
});

export const createRequestSuccess = <P, T extends string>(type: T) => (
  payload: P,
  hash: string
) => ({
  type,
  payload,
  meta: {
    hash,
    receivedAt: Date.now(),
  },
});

export const createRequestPending = <T extends string>(type: T) => (hash: string) => ({
  type,
  meta: {
    hash,
  },
});
