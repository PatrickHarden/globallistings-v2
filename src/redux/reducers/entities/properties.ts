import { Dictionary } from '../../../types/common/dictionary';
import { Property } from '../../../types/property';
import * as actions from '../../actions/searchProperties';
import { ActionType } from 'typesafe-actions';

export type Actions = ActionType<typeof actions>;

export default (state: Dictionary<Property> = {}, action: Actions): Dictionary<Property> => {
  switch (action.type) {
    case 'SEARCH_PROPERTIES_SUCCESS':
      return {
        ...state,
        ...action.payload.entities.properties,
      };
    case 'SEARCH_PROPERTY_SUCCESS':
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};
