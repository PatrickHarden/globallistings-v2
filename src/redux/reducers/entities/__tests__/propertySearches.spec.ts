import reducer from '../propertySearches';
import { DeepPartial } from '../../../../types/common/deep-partial';
import { NormalisedSearchResult } from '../../../../types/results.js';
import { Actions } from '../properties';

const MOCK_HASH = 'MOCK_HASH';
const CREATED_AT_MOCK = 123123123;

describe('handle SUCCESS action', () => {
  const p1: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P1: {
          numberOfBedrooms: 1,
        },
      },
    },
    result: ['P1'],
  };

  const p2: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P2: {
          numberOfBedrooms: 2,
        },
      },
    },
    result: ['P2'],
  };

  const p1Modified: DeepPartial<NormalisedSearchResult> = {
    entities: {
      properties: {
        P1: {
          numberOfBedrooms: 3,
        },
      },
    },
    result: ['P1'],
  };

  it('should correctly handle single SUCCESS', () => {
    const successAction: Actions = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    };
    const newState = reducer({}, successAction);

    expect(newState).toMatchSnapshot();
  });

  it('should overwrite old properties with new', () => {
    const successAction1: Actions = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    };
    const successAction2: Actions = {
      payload: p2 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    };
    const newState1 = reducer({}, successAction1);
    const newState2 = reducer(newState1, successAction2);

    expect(newState2).toMatchSnapshot();
  });

  it('should overwrite old properties with same id', () => {
    const successAction1: Actions = {
      payload: p1 as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    };
    const successAction2: Actions = {
      payload: p1Modified as NormalisedSearchResult,
      type: 'SEARCH_PROPERTIES_SUCCESS',
      meta: { hash: MOCK_HASH, receivedAt: CREATED_AT_MOCK },
    };
    const newState1 = reducer({}, successAction1);
    const newState2 = reducer(newState1, successAction2);

    expect(newState2).toMatchSnapshot();
  });
});
