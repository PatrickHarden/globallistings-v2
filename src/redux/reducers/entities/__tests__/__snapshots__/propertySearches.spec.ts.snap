// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle SUCCESS action should correctly handle single SUCCESS 1`] = `
Object {
  "MOCK_HASH": Array [
    "P1",
  ],
}
`;

exports[`handle SUCCESS action should overwrite old properties with new 1`] = `
Object {
  "MOCK_HASH": Array [
    "P2",
  ],
}
`;

exports[`handle SUCCESS action should overwrite old properties with same id 1`] = `
Object {
  "MOCK_HASH": Array [
    "P1",
  ],
}
`;
